File: src/Go/speed_test/unit_allocator.go
package speed_test

import (
	. "../common/nanotime"
	"time"
)

const UnitSize = 24
const StepSize = 50

type UnitAllocator struct {
	RunDuration time.Duration
	Last        *AllocationUnit
	// Statistics
	AllocationCount int64
}

type AllocationUnit struct {
	Field1 int64
	Field2 int64
	Field3 int64
}

func NewUnitAllocator(runDuration time.Duration) *UnitAllocator {
	a := &UnitAllocator{
		RunDuration: runDuration,
	}
	return a
}

func (a *UnitAllocator) Run() {
	lastTimestamp := Nanotime().Nanoseconds()
	endTimestamp := lastTimestamp + a.RunDuration.Nanoseconds()
	last := &AllocationUnit{}

	for lastTimestamp < endTimestamp {
		for i := 0; i < StepSize; i++ {
			last = new(AllocationUnit) // Just to make sure there is a heap allocation
			// last = &AllocationUnit{} // This version runs with the same speed
		}
		a.AllocationCount += StepSize
		lastTimestamp = Nanotime().Nanoseconds()
	}
	a.Last = last // To suppress unused var error; changes nothing
}

File: src/CSharp/Main/SpeedTest/UnitAllocator.cs
using System;
using System.Diagnostics;
using Benchmarking.Common;

namespace GCBurn.SpeedTest
{
    public class UnitAllocator : IActivity
    {
#pragma warning disable 169
        public class Unit { long Field1, Field2; }
#pragma warning restore 169

        public const long UnitSize = 2 * sizeof(long);
        public const int StepSize = 50;
        public TimeSpan RunDuration = TimeSpan.Zero;
        public long AllocationCount = 0;
        private object _sink;

        public UnitAllocator(TimeSpan runDuration) => RunDuration = runDuration;

        public void Run()
        {
            var duration = RunDuration;
            var lastTimestamp = Stopwatch.GetTimestamp();
            var endTimestamp = lastTimestamp + (long) (duration.TotalSeconds * Stopwatch.Frequency);
            var allocationCount = 0;
            object last = null;
            while (lastTimestamp < endTimestamp) {
                for (var i = 0; i < StepSize; i++)
                    last = new Unit();
                allocationCount += StepSize;
                lastTimestamp = Stopwatch.GetTimestamp();
            }            
            _sink = last; // Just to make sure JIT / compiler somehow doesn't eliminate the allocation at all
            AllocationCount += allocationCount;
        }
    }
}
